### PR TITLE
Rock, Scissors, Paper problem solution using 3D DP in Java  #1591

### LINKED ISSUE #1591: Rock, Scissors, Paper problem using 3D DP in Java
I wouls like to contribute to this problem.

### PR DESCRIPTION
**Pull request by shreyasrivastava936**
Fixes #1591

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Input is taken dynamically.
- [x] Sample Input / Output is added at the end of file.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Changes proposed in this pull request:
- applied dp to solve problem for rock, Scissor, Paper problem.
- 

#### Language Used:
- Java
- 

#### Files Added:
- RSP.java
- 

> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!

### CODE REVIEW

**Comment by shreyasrivastava936** [comment]
ping @singh-shreya6  @Vaibhavraj10 @MastersAbh

**Comment by shreyasrivastava936** [comment]
@jainaman224 please check it once .

**Comment by SakshiShreya** [comment]
mention the issue number after Fixes #.

File: RSP.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        
+        res+=probpaper(r-1, p, s)*(((double)(r*p))/total);
+        res+=probpaper(r, p-1, s)*(((double)(s*p))/total);
+        res+=probpaper(r, p, s-1)*(((double)(r*s))/total);
+        
+        return dp[r][p][s] = res;
+    }
+    
+     public static double probscissor(int r, int p, int s)
+    {
```
Comment (jainaman224) [inline RSP.java]: Incorrect indentation.

File: RSP.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            return 1.0;
+            
+        if(dp[r][p][s] != -1.0)
+            return dp[r][p][s];
+        
+        //total no. of fight = rC1.pC1 + pC1.sC1 + sC1.rC1
+        double total = r*p + p*s + s*r;
+        double res = 0.0;
+        
+        res+=probpaper(r-1, p, s)*(((double)(r*p))/total);
```
Comment (jainaman224) [inline RSP.java]: Add spaces around operator like `+=`, `+`,  `=`.

File: RSP.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	        for(int j=1;j<=paper;j++)
+	        {
+	            for(int k=0; k<=scissor; k++)
+	            {
+	                dp[i][j][k] = -1.0;
+	            }
+	        }
+	    }
+	    
+	}
```
Comment (jainaman224) [inline RSP.java]: Use spaces instead of tabs.

File: RSP.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
+import java.util.Scanner;
```
Comment (jainaman224) [inline RSP.java]: Add this file to a folder.
folder name should be Rock_Paper_Scissor
And file name should be Rock_Paper_Scissor.java

**Comment by shreyasrivastava936** [comment]
@jainaman224 I've changed the file .check it once

**Comment by shreyasrivastava936** [comment]
issue no. #1591

File: Rock_Paper_Scissor/Rock_Paper_Scissor.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        
+        res += probscissor(r-1, p, s)*(((double)(r*p))/total);
+        res += probscissor(r, p-1, s)*(((double)(s*p))/total);
+        res += probscissor(r, p, s-1)*(((double)(r*s))/total);
+        
+        return dp[r][p][s] = res;
+    }
+	
+    public static void main(String[] args) {
+	    int rock, scissor, paper;
```
Comment (jainaman224) [inline Rock_Paper_Scissor/Rock_Paper_Scissor.java]: Use 4 space indentation.

File: Rock_Paper_Scissor/Rock_Paper_Scissor.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+	    initializeDp(rock, paper, scissor);
+	    double ansPaper = probpaper(rock, paper, scissor);
+	    
+	    initializeDp(rock, paper, scissor);
+	    double ansScissor = probscissor(rock, paper, scissor);
+	    
+	    System.out.println("Probability of rock to survive is " + ansRock);
+	    System.out.println("Probability of scissor to survive is " + ansScissor);
+	    System.out.println("Probability of paper to survive is " + ansPaper);    
+	}
```
Comment (jainaman224) [inline Rock_Paper_Scissor/Rock_Paper_Scissor.java]: Inconsistent indentation.

File: Rock_Paper_Scissor/Rock_Paper_Scissor.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+   public static void initializeDp(int rock, int paper, int scissor) {
+    	for(int i=1;i<=rock;i++) {
```
Comment (jainaman224) [inline Rock_Paper_Scissor/Rock_Paper_Scissor.java]: Space around operators like `==`, `=`, `+`, `-`, `<=`.
Space after operator like `;`, `,`.

File: Rock_Paper_Scissor/Rock_Paper_Scissor.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+	   for(int j=1;j<=paper;j++) {
+	      for(int k=0; k<=scissor; k++) {
+		  dp[i][j][k] = -1.0;
+		    }
+	        }
+	    }   
+	}
+}
```
Comment (jainaman224) [inline Rock_Paper_Scissor/Rock_Paper_Scissor.java]: Add sample input and output.

**Comment by shreyasrivastava936** [comment]
@jainaman224 done with indentation problem. kindly check the file

**Comment by shreyasrivastava936** [comment]
@jainaman224 check the commits please

File: Rock_Paper_Scissor/Rock_Paper_Scissor.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+    res += probscissor (r - 1, p, s) * (((double) (r * p)) / total);
+    res += probscissor (r, p - 1, s) * (((double) (s * p)) / total);
+    res += probscissor (r, p, s - 1) * (((double) (r * s)) / total);
+
+    return dp[r][p][s] = res;
+  }
+
+  public static void main (String[]args)
+  {
+    int rock, scissor, paper;
```
Comment (somya-kapoor) [inline Rock_Paper_Scissor/Rock_Paper_Scissor.java]: Please correct the indentation(4 space indentation)

File: Rock_Paper_Scissor/Rock_Paper_Scissor.java [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+      res += probrock (r - 1, p, s) * (((double) (r * p)) / total);	//prob. of fight btw r&p = r*p/total
+      res += probrock (r, p - 1, s) * (((double) (s * p)) / total);	//prob. of fight btw s&p = s*p/total
+      res += probrock (r, p, s - 1) * (((double) (r * s)) / total);	//prob. of fight btw r&s= r*s/total 
+
+      return dp[r][p][s] = res;
+  }
+
+  public static double probpaper (int r, int p, int s)
+  {
+    if (p == 0 || r == 0)
```
Comment (somya-kapoor) [inline Rock_Paper_Scissor/Rock_Paper_Scissor.java]: Please correct the indentation(4 space indentation)

File: Rock_Paper_Scissor/Rock_Paper_Scissor.java [hunk 11]
```diff
@@ -0,0 +1,123 @@
+import java.util.Scanner;
+public class Main
```
Comment (somya-kapoor) [inline Rock_Paper_Scissor/Rock_Paper_Scissor.java]: Leave a line before main class

**Comment by somya-kapoor** [comment]
Please correct the indentation(4 space indentation)

**Comment by somya-kapoor** [comment]
@shreyasrivastava936 Are you working on it?

**Comment by shreyasrivastava936** [comment]
Yes

**Comment by shreyasrivastava936** [comment]
@somya-kapoor  I've made the changes .please review the file

File: Rock_Paper_Scissor/Rock_Paper_Scissor.java [hunk 12]
```diff
@@ -0,0 +1,118 @@
+import java.util.Scanner;
+
+public class Main
+{
+	public static int large = 500;
+    public static double dp[][][] = new double[large][large][large];
+
+    public static double probrock (int r, int p, int s) {
+	    if (s == 0 || r == 0)
+		  return 0.0;
```
Comment (somya-kapoor) [inline Rock_Paper_Scissor/Rock_Paper_Scissor.java]: inconsistency in indentation

File: Rock_Paper_Scissor/Rock_Paper_Scissor.java [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+
+		if (p == 0)
+		  return 1.0;
```
Comment (somya-kapoor) [inline Rock_Paper_Scissor/Rock_Paper_Scissor.java]: inconsistency in indentation
use 4 space indentation

File: Rock_Paper_Scissor/Rock_Paper_Scissor.java [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+
+		if (dp[r][p][s] != -1.0)
+		  return dp[r][p][s];
```
Comment (somya-kapoor) [inline Rock_Paper_Scissor/Rock_Paper_Scissor.java]: inconsistency in indentation
use 4 space indentation

**Comment by somya-kapoor** [comment]
@shreyasrivastava936 Inconsistency in indentation
Please use 4 space indentation

File: Rock_Paper_Scissor/Rock_Paper_Scissor.java (hunk 9 above)
Comment (shreyasrivastava936) [reply Rock_Paper_Scissor/Rock_Paper_Scissor.java]: @somya-kapoor I've made the changes, please check the file

**Comment by shreyasrivastava936** [comment]
@somya-kapoor please review  the changes

File: Rock_Paper_Scissor/Rock_Paper_Scissor.java [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+    res += probrock (r - 1, p, s) * (((double) (r * p)) / total);	//prob. of fight btw r&p = r*p/total
+    res += probrock (r, p - 1, s) * (((double) (s * p)) / total);	//prob. of fight btw s&p = s*p/total
+    res += probrock (r, p, s - 1) * (((double) (r * s)) / total);	//prob. of fight btw r&s= r*s/total 
+
+    return dp[r][p][s] = res;
+}
+
+public static double probpaper (int r, int p, int s) {
+    if (p == 0 || r == 0)
+    return 0.0;
```
Comment (somya-kapoor) [inline Rock_Paper_Scissor/Rock_Paper_Scissor.java]: Indentation of return statement

File: Rock_Paper_Scissor/Rock_Paper_Scissor.java [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+
+    if (s == 0)
+    return 1.0;
```
Comment (somya-kapoor) [inline Rock_Paper_Scissor/Rock_Paper_Scissor.java]: Indentation of return statement

File: Rock_Paper_Scissor/Rock_Paper_Scissor.java [hunk 17]
```diff
@@ -0,0 +1,118 @@
+import java.util.Scanner;
+
+public class Main
+{
+    public static int large = 500;
+    public static double dp[][][] = new double[large][large][large];
+
+    public static double probrock (int r, int p, int s) {
+    if (s == 0 || r == 0)
+    return 0.0;
```
Comment (somya-kapoor) [inline Rock_Paper_Scissor/Rock_Paper_Scissor.java]: Indentation of return statement

File: Rock_Paper_Scissor/Rock_Paper_Scissor.java [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+
+    if (p == 0)
+    return 1.0;
```
Comment (somya-kapoor) [inline Rock_Paper_Scissor/Rock_Paper_Scissor.java]: Indentation of return statement

File: Rock_Paper_Scissor/Rock_Paper_Scissor.java [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+
+    if (dp[r][p][s] != -1.0)
+    return dp[r][p][s];
```
Comment (somya-kapoor) [inline Rock_Paper_Scissor/Rock_Paper_Scissor.java]: Indentation of return statement

File: Rock_Paper_Scissor/Rock_Paper_Scissor.java [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+
+public static double probpaper (int r, int p, int s) {
+    if (p == 0 || r == 0)
+    return 0.0;
+
+    if (s == 0)
+    return 1.0;
+
+    if (dp[r][p][s] != -1.0)
+    return dp[r][p][s];
```
Comment (somya-kapoor) [inline Rock_Paper_Scissor/Rock_Paper_Scissor.java]: Indentation of return statement

File: Rock_Paper_Scissor/Rock_Paper_Scissor.java [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+    res += probpaper (r - 1, p, s) * (((double) (r * p)) / total);
+    res += probpaper (r, p - 1, s) * (((double) (s * p)) / total);
+    res += probpaper (r, p, s - 1) * (((double) (r * s)) / total);
+
+    return dp[r][p][s] = res;
+}
+
+public static double probscissor (int r, int p, int s) {
+    if (p == 0 || s == 0)
+    return 0.0;
```
Comment (somya-kapoor) [inline Rock_Paper_Scissor/Rock_Paper_Scissor.java]: Indentation of return statement. Fix in the whole file

**Comment by shreyasrivastava936** [comment]
@somya-kapoor please recheck the file. Hope all the things are fixed now

**Comment by somya-kapoor** [comment]
@shreyasrivastava936 Please squash the commits

**Comment by somya-kapoor** [comment]
@shreyasrivastava936 Are you working on it?

**Comment by shreyasrivastava936** [comment]
yes

**Comment by somya-kapoor** [comment]
@shreyasrivastava936 Due to no activity for more than 7 days, I am closing this issue.

**Pull request closed without merging**